keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,117 @@
+
+## Motivation
+
+The main problem with slow starting containers is that you have to leave them enough time to start before having `livenessProbe` fail `failureThreshold` times (otherwise `kubelet` kills them before they have a chance to be up).
+This is even worse when load on the nodes can increase this long startup time since you have to take into account the worse case scenario (let's call it **T**).
+
+You can handle that today with various strategies:
+
+- set `initialDelaySeconds` > **T**: the `livenessProbe` isn't run for **T** time (so you're blind in the monitoring), you have to wait at least **T** until `kubelet` kills your container after a deadlock - this could only happen at the beginning of the container lifetime
+- set `failureThreshold` * `periodSeconds` > **T**: the `livenessProbe` is run every `periodSeconds` (so you can monitor the probe status), **but you still** have to wait at least **T** until `kubelet` kills your container after a deadlock - this could happen at any time
+
+### Goals
+
+Add an elegant and self-explanatory solution to running slow starting containers in Kubernetes with health probes enabled.
+
+By using `maxInitialFailureCount` * `periodSeconds` > **T** you have the following benefits:
+
+- `livenessProbe` is run every `periodSeconds` - good for monitoring
+- container has enough time to start in the worse case scenario
+- when `livenessProbe` has succeeded once, it has to fail only `failureThreshold` (should be set to 1 or at least low) before `kubelet` kills it
+
+### Non-Goals
+
+Do not debate on whether end-users should or not deploy such containers in Kubernetes.
+
+We all have legacy projects and not enough budget to rewrite them over clound native concerns.
+
+## Proposal
+
+### Implementation Details
+
+Adding a boolean `hasInitialized` to the kubelet worker keeping track of the state of the container (did the probe ever succeed?) and use it as a new condition to return `true` to the probe state in `worker.go`.
+
+You can see the implementation in PR [#71449].
+
+[#71449]: https://github.com/kubernetes/kubernetes/pull/71449
+
+### Risks and Mitigations
+
+`kubelet` handles container/pod lifecycle-related functions by relying on the underlying container runtime to persist the states. For probes and/or lifecycle hooks, kubelet rely on in-memory states only.
+
+If the `kubelet` is restarted, and `maxInitialFailureCount` is set very high, the effect would be the same as having previously a high `failureThreshold` potentially making the container downtime much longer if a dead-lock occurs during initialization.
+
+Documentation must be clearly stating that this feature should only be used in particular cases.
+
+## Design Details
+
+### Test Plan
+
+The following test cases can be covered by calling the `fakeExecProber` a number of times to verify:
```

Comment:
I think there should also be a case of container probe succeeds before the `maxInitialFailureCount` is up and then fails for `failureThreshold` but the total time to failure is less than `maxInitialFailureCount` -- that's the whole point of this.

Reply:
This case is addressed by the fact that `maxInitialFailureCount` isn't checked once the probe has succeeded once, which is remembered by the state `hasInitialized`.

Reply:
I understand, but that's something you should unit test for.  This test should run very quickly.

Reply:
ok